File: tests/reboot/zod/methods/servicer_api.ts
import {
  EMPTY_ARRAY,
  EMPTY_RECORD,
  reader,
  transaction,
  writer,
} from "@reboot-dev/reboot-api";
import { z } from "zod/v4";

export const LiteralValueType = z.literal(["option1", "option2", "option3"]);

export const Test = {
  state: {
    data: z.string().meta({ tag: 1 }),
    literalValue: LiteralValueType.meta({ tag: 2 }),

    // Empty default values.
    stringDefaultValue: z.string().default("").meta({ tag: 3 }),
    booleanDefaultValue: z.boolean().default(false).meta({ tag: 4 }),
    numberDefaultValue: z.number().default(0).meta({ tag: 5 }),
    arrayDefaultValue: z
      .array(z.string())
      .default(EMPTY_ARRAY)
      .meta({ tag: 6 }),
    recordDefaultValue: z
      .record(z.string(), z.string())
      .default(EMPTY_RECORD)
      .meta({ tag: 7 }),
  },
  methods: {
    construct: writer({
      factory: {},
      request: {},
      response: z.void(),
    }),
    writer: writer({
      request: {
        data: z.string().meta({ tag: 1 }),
        previousLiteralValue: LiteralValueType.meta({ tag: 2 }),
        newLiteralValue: LiteralValueType.optional().meta({ tag: 3 }),
      },
      response: z.void(),
    }),
    checkDefaults: reader({ request: {}, response: z.void() }),
    reader: reader({
      request: {},
      response: {
        data: z.string().meta({ tag: 1 }),
        literalValue: LiteralValueType.meta({ tag: 2 }),
      },
    }),
    transaction: transaction({
      request: {},
      response: z.void(),
    }),
    transactionReader: reader({
      request: {},
      response: z.void(),
    }),
    transactionWriter: writer({
      request: {
        shouldFail: z.boolean().meta({ tag: 1 }),
      },
      response: z.void(),
    }),
  },
};

export const api = { Test };

File: tests/reboot/pydantic_web/optional_fields/servicer_api.py
from reboot.api import API, Field, Methods, Model, Reader, Type, Writer
from typing import Optional


class State(Model):
    message: Optional[str] = Field(tag=1, default=None)
    counter: Optional[int] = Field(tag=2, default=None)


class SetMessageAndCounterRequest(Model):
    message: Optional[str] = Field(tag=1)
    counter: Optional[int] = Field(tag=2)


class GetStateResponse(Model):
    message: Optional[str] = Field(tag=1)
    counter: Optional[int] = Field(tag=2)


TestMethods = Methods(
    set_message_and_counter=Writer(
        request=SetMessageAndCounterRequest,
        response=None,
        mcp=None,
    ),
    get_state=Reader(
        request=None,
        response=GetStateResponse,
        mcp=None,
    ),
)

api = API(
    Test=Type(
        state=State,
        methods=TestMethods,
    ),
)
